### PR TITLE
FR: Add spec validators for the WRP messages

### PR DESCRIPTION
**Pull request by denopink**
## Overview

related to #25, #78, xmidt-org/scytale#88, xmidt-org/talaria#153 and builds on top of #80.

### tl;dr
This pr introduces our wrp `spec validators` built with our validation framework introduced in #80 which only validates the opinionated portions of the spec. Clients can leverage these prebuilt validators to validate their messages.

<details>
<summary> Explanation</summary>

* Validates requirements described at https://xmidt.io/docs/wrp/basics/#overarching-guidelines
* Validates requirements described at https://xmidt.io/docs/wrp/basics/#locators
* Validates message type values
* Does not cover validators for a specific message type (future prs)

Clients can leverage our prebuilt validators to validate their messages. One set of these validators are our `spec validators`:
```go
msgv, err := NewTypeValidator(
	// Validates found msg types
	map[MessageType]Validator{
		// Validates opinionated portions of the spec
		SimpleEventMessageType: SpecValidators,
		// Only validates Source and nothing else
		SimpleRequestResponseMessageType: SourceValidator,
	},
	// Validates unfound msg types
	AlwaysInvalid)
if err != nil {
	return
}

foundErrSuccess1 := msgv.Validate(Message{
	Type:        SimpleEventMessageType,
	Source:      "MAC:11:22:33:44:55:66",
	Destination: "MAC:11:22:33:44:55:61",
}) // Found success
foundErrSuccess2 := msgv.Validate(Message{
	Type:        SimpleRequestResponseMessageType,
	Source:      "MAC:11:22:33:44:55:66",
	Destination: "invalid:a-BB-44-55",
}) // Found success
foundErrFailure := msgv.Validate(Message{
	Type:        Invalid0MessageType,
	Source:      "invalid:a-BB-44-55",
	Destination: "invalid:a-BB-44-55",
}) // Found error
unfoundErrFailure := msgv.Validate(Message{Type: CreateMessageType}) // Unfound error
fmt.Println(foundErrSuccess1 == nil, foundErrSuccess2 == nil, foundErrFailure == nil, unfoundErrFailure == nil)
// Output: true true false false
```

</details>

<details>
<summary>Type of Change(s)</summary>

- Non-breaking Enhancement
- All new and existing tests passed.

</details>

### CODE REVIEW

File: spec_validator.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+// SpecValidators is a WRP validator that ensures messages are valid based on
+// each spec validator in the list. Only validates the opinionated portions of the spec.
+// SpecValidators validates the following fields: UTF8 (all string fields), MessageType, Source, Destination
+func SpecValidators() Validators {
+	return Validators{UTF8Validator, MessageTypeValidator, SourceValidator, DestinationValidator}
+}
+
+// UTF8Validator is a WRP validator that takes messages and validates that it contains UTF-8 strings.
+var UTF8Validator ValidatorFunc = func(m Message) error {
```
Comment (kristinapathak) [inline spec_validator.go]: Please make all of these `var ValidatorFunc`s into `func`s like `SpecValidators()` so they can't be modified by consumers.  Then move `locatorPattern` to the top so all package level vars are listed before any functions.  Then the general file structure will match the rest of our codebase.

File: spec_validator.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+	return nil
+}
+
+// SourceValidator is a WRP validator that takes messages and validates their Source.
+// Only mac and uuid sources are validated. Serial, event and dns sources are
+// not validated.
+var SourceValidator ValidatorFunc = func(m Message) error {
+	if err := validateLocator(m.Source); err != nil {
+		return fmt.Errorf("%w: %v", ErrorInvalidSource, err)
```
Comment (kristinapathak) [inline spec_validator.go]: I don't think the source is sensitive information so we might as well include them in the error for any human who reads it. 🙂  The same can be done when validating the destination.
```suggestion
		return fmt.Errorf("%w '%s': %v", ErrorInvalidSource, m.Source, err)
```

File: spec_validator.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	`^(?P<scheme>(?i)` + macPrefix + `|` + uuidPrefix + `|` + eventPrefix + `|` + dnsPrefix + `|` + serialPrefix + `):(?P<authority>[^/]+)?`,
+)
+
+// validateLocator validates a given locator's scheme and authority (ID).
+// Only mac and uuid schemes' IDs are validated. IDs from serial, event and dns schemes are
+// not validated.
+func validateLocator(l string) error {
+	match := locatorPattern.FindStringSubmatch(l)
+	if match == nil {
+		return fmt.Errorf("spec scheme not found")
```
Comment (kristinapathak) [inline spec_validator.go]: spec scheme not found is a bit confusing (at least for me) - what about something like `value given doesn't match expected locator pattern`?  Then the error would be wrapped as...
```
invalid Source name 'abcd': value given doesn't match expected locator pattern
```

File: spec_validator.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+// locatorPattern is the precompiled regular expression that all source and dest locators must match.
+// Matching is partial, as everything after the authority (ID) is ignored. https://xmidt.io/docs/wrp/basics/#locators
+var locatorPattern = regexp.MustCompile(
+	`^(?P<scheme>(?i)` + macPrefix + `|` + uuidPrefix + `|` + eventPrefix + `|` + dnsPrefix + `|` + serialPrefix + `):(?P<authority>[^/]+)?`,
+)
+
+// validateLocator validates a given locator's scheme and authority (ID).
+// Only mac and uuid schemes' IDs are validated. IDs from serial, event and dns schemes are
+// not validated.
+func validateLocator(l string) error {
```
Comment (kristinapathak) [inline spec_validator.go:119]: I would make all of the errors in this function have some unexported error vars that we can check against in the tests.  Then the code kind of documents itself, and testing is easy.  In places where we're returning a constant string error currently, the var can be returned instead without any `fmt.Errorf` calls in the function.

File: spec_validator.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+		)
+
+		if invalidCharacter != -1 {
+			return fmt.Errorf("invalid character %v", strconv.QuoteRune(invalidCharacter))
+		} else if len(idPart) != macLength {
+			return errors.New("invalid mac length")
+		}
+	case uuidPrefix:
+		if _, err := uuid.Parse(idPart); err != nil {
+			return err
```
Comment (kristinapathak) [inline spec_validator.go]: I would wrap the error here - then we have a more informative error and in our tests we can easily verify that the function failed where we expected it to:
```suggestion
			return fmt.Errrof("%w: %v", errInvalidUUID, err)
```

at top of file:
```
    var errInvalidUUID = errors.New("invalid UUID authority")

**Review by kristinapathak — COMMENTED** [review]
Great start! The code itself looks good, mostly some structure and error suggestions.

**Review by kristinapathak — APPROVED** [review]
💯 Looks great!
